Add Acl whitelist Accounts + Contracts

gRoussac opened the pull request:
This PR intends to add a common whitelist called ACL handling account / contract  as Key whitelisting for the mint entry point + tests

gRoussac committed:
Add MintingMode::Acl

gRoussac committed:
Wording

gRoussac committed:
Revert change

gRoussac committed:
Add some tests

gRoussac committed:
linting

gRoussac committed:
small change

gRoussac committed:
move migrating white list to utils

gRoussac committed:
Modify get_verified_caller and acl whitelist to handle whitelisting package

gRoussac committed:
naming tests

gRoussac committed:
small change

gRoussac committed:
Update readme + add tests

gRoussac committed:
Test CI/CD

gRoussac commented on contract/Cargo.toml:
```diff
@@ -7,7 +7,7 @@ edition = "2018"
   "test-support",
 ], optional = true }
-casper-types = "1.5"
+casper-types = "=1.5"
```
Please notice that change for CI/CD

gRoussac committed:
Add test

gRoussac commented:
@darthsiroftardis @deuszex  Please note that [this logic](https://github.com/casper-ecosystem/cep-78-enhanced-nft/blob/dev/contract/src/main.rs#L537-L540) is being changed and that you can not install those mods anymore now 
MintingMode::Installer+NFTHolderMode:Contracts 

typically this request will not install anymore

```
        .with_holder_mode(NFTHolderMode::Contracts)
        .with_ownership_mode(OwnershipMode::Minter)
        .with_reporting_mode(OwnerReverseLookupMode::NoLookUp)
        .with_minting_mode(MintingMode::Installer)
        .with_whitelist_mode(WhitelistMode::Locked)
        .with_contract_whitelist(contract_whitelist)
```

as `MintingMode::Installer` is a private mode as per comment sugests https://github.com/casper-ecosystem/cep-78-enhanced-nft/blob/dev/contract/src/modalities.rs#L59 

it should be 

```
        .with_minting_mode(MintingMode::Acl)
```

gRoussac committed:
small change

ACStone-MTS commented on README.md:
```diff
@@ -97,28 +94,13 @@ This modality is an optional installation parameter and will default to the `Mix
 
 1. `Installer`: This mode restricts the ability to mint new NFT tokens only to the installing account of the NFT contract.
 2. `Public`: This mode allows any account to mint NFT tokens.
+3. `Acl`: This mode allows whitelisted account/contract/contract from a package to mint NFT tokens.
```
```suggestion
3. `Acl`: This mode allows whitelisted accounts, contracts or contracts from a package to mint NFT tokens.
```

ACStone-MTS replied:
Is Acl an acronym? If so, it should be written as ACL.

gRoussac replied:
Renamed to ACL thanks

ACStone-MTS commented on README.md:
```diff
@@ -127,6 +109,26 @@ mode cannot be changed once the contract has been installed. The mode is set by
+
+#### WhitelistMode
+
+The `WhitelistMode` dictates if the Acl whitelist restricting access to the mint entrypoint can be updated. There are currently
```
```suggestion
The `WhitelistMode` dictates if the Acl whitelist restricting access to the mint entry point can be updated. There are currently two options:
```

ACStone-MTS commented on README.md:
```diff
@@ -127,6 +109,26 @@ mode cannot be changed once the contract has been installed. The mode is set by
+#### WhitelistMode
+
+The `WhitelistMode` dictates if the Acl whitelist restricting access to the mint entrypoint can be updated. There are currently
+two options:
```
```suggestion
```

ACStone-MTS commented on README.md:
```diff
@@ -127,6 +109,26 @@ mode cannot be changed once the contract has been installed. The mode is set by
+1. `Unlocked`: The Acl whitelist is unlocked and can be updated via the set variables endpoint.
+2. `Locked`: The Acl whitelist is locked and cannot be updated further.
+
+If the `WhitelistMode` is set to `Locked` an Acl whitelist of entity keys must be provided on installation. This whitelist dictates which
```
```suggestion
If the `WhitelistMode` is set to `Locked` an Acl whitelist of entity keys must be provided on installation. This whitelist dictates which entities can mint NFTs in the restricted `Acl` minting mode. These entities include `Accounts`, `Contracts` and `Contracts from a package`.
```

ACStone-MTS commented on README.md:
```diff
@@ -127,6 +109,26 @@ mode cannot be changed once the contract has been installed. The mode is set by
+2. `Locked`: The Acl whitelist is locked and cannot be updated further.
+
+If the `WhitelistMode` is set to `Locked` an Acl whitelist of entity keys must be provided on installation. This whitelist dictates which
+entity either `Account` / `Contract` / `Contract from a package` are allowed to mint NFTs in the restricted `Acl` minting mode.
```
```suggestion
```

These are showing up as separate lines, which is strange for a README and may cause formatting issues.

ACStone-MTS commented on README.md:
```diff
@@ -127,6 +109,26 @@ mode cannot be changed once the contract has been installed. The mode is set by
+If the `WhitelistMode` is set to `Locked` an Acl whitelist of entity keys must be provided on installation. This whitelist dictates which
+entity either `Account` / `Contract` / `Contract from a package` are allowed to mint NFTs in the restricted `Acl` minting mode.
+
+This `WhitelistMode` is an optional installation parameter and will be set to unlocked if not passed. However, the whitelist mode itself
```
```suggestion
This `WhitelistMode` is an optional installation parameter and will be set to unlocked if not passed. However, the whitelist mode itself cannot be changed once the contract has been installed. The mode is passed in as a `u8` value to `whitelist_mode` runtime argument.
```

ACStone-MTS commented on README.md:
```diff
@@ -127,6 +109,26 @@ mode cannot be changed once the contract has been installed. The mode is set by
+entity either `Account` / `Contract` / `Contract from a package` are allowed to mint NFTs in the restricted `Acl` minting mode.
+
+This `WhitelistMode` is an optional installation parameter and will be set to unlocked if not passed. However, the whitelist mode itself
+cannot be changed once the contract has been installed. The mode is passed in as a `u8` value to `whitelist_mode` runtime argument.
```
```suggestion
```

ACStone-MTS commented on README.md:
```diff
@@ -381,9 +383,9 @@ The following are the optional parameters that can be passed in at the time of i
+- `"whitelist_mode"`: The [`WhitelistMode`](#whitelistmode) modality dictates whether the acl whitelist can be updated. This optional parameter will default to an unlocked whitelist that can be updated post installation. This parameter cannot be changed once the contract has been installed.
 - `"holder_mode"`: The [`NFTHolderMode`](#nftholdermode) modality dictates which entities can hold NFTs. This is an optional parameter and will default to a mixed mode allowing either `Accounts` or `Contracts` to hold NFTs. This parameter cannot be changed once the contract has been installed.
-- `"contract_whitelist"`: The contract whitelist is a list of contract hashes that specifies which contracts can call the `mint()` entrypoint to mint NFTs. This is an optional parameter which will default to an empty whitelist. This value can be changed via the `set_variables` post installation. If the whitelist mode is set to locked, a non-empty whitelist must be passed; else, installation of th ...
+- `"acl_whitelist"`: The acl whitelist is a list of account / contract hashes / package that specifies which entity can call the `mint()` entrypoint to mint NFTs. This is an optional parameter which will default to an empty whitelist. This value can be changed via the `set_variables` post installation. If the whitelist mode is set to locked, a non-empty whitelist must be passed; else, installatio ...
```
```suggestion
- `"acl_whitelist"`: The acl whitelist is a list of account / contract hashes / packages that specifies which entity can call the `mint()` entrypoint to mint NFTs. This is an optional parameter which will default to an empty whitelist. This value can be changed via the `set_variables` post installation. If the whitelist mode is set to locked, a non-empty whitelist must be passed. If the whitelist mode is set to locked and you do not provide a non-empty whitelist, the contract will fail.
```

gRoussac replied:
@ACStoneCL thanks for the good watch

gRoussac committed:
Naming ACL

gRoussac committed:
Update README.md

Co-authored-by: Adam Stone <97986246+ACStoneCL@users.noreply.github.com>

gRoussac committed:
Update README.md

Co-authored-by: Adam Stone <97986246+ACStoneCL@users.noreply.github.com>

gRoussac committed:
Update README.md

Co-authored-by: Adam Stone <97986246+ACStoneCL@users.noreply.github.com>

gRoussac committed:
Update README.md

Co-authored-by: Adam Stone <97986246+ACStoneCL@users.noreply.github.com>

gRoussac committed:
Update README.md

Co-authored-by: Adam Stone <97986246+ACStoneCL@users.noreply.github.com>

gRoussac committed:
Update README.md

Co-authored-by: Adam Stone <97986246+ACStoneCL@users.noreply.github.com>

gRoussac committed:
Update README.md

Co-authored-by: Adam Stone <97986246+ACStoneCL@users.noreply.github.com>

gRoussac committed:
Update README.md

Co-authored-by: Adam Stone <97986246+ACStoneCL@users.noreply.github.com>

gRoussac committed:
Naming

gRoussac committed:
Small change

gyroflaw reviewed (commented):
overall looks amazing 👍, 
and I think some parts can be refactored

gRoussac commented on contract/src/main.rs:
```diff
@@ -568,16 +586,32 @@ pub extern "C" fn mint() {
-        };
+    // Revert if minting is acl and caller is not whitelisted.
+    if MintingMode::ACL == minting_mode {
+        let is_whitelisted = utils::get_dictionary_value_from_key::<bool>(
+            ACL_WHITELIST,
+            &utils::encode_dictionary_item_key(caller),
+        )
+        .unwrap_or_default();
+        if !is_whitelisted {
+            match caller.tag() {
+                KeyTag::Hash => {
+                    if let Some(contract_package) = contract_package {
+                        let is_package_whitelisted = utils::get_dictionary_value_from_key::<bool>(
+                            ACL_WHITELIST,
+                            &utils::encode_dictionary_item_key(contract_package),
+                        )
+                        .unwrap_or_default();
+                        if !is_package_whitelisted {
+                            runtime::revert(NFTCoreError::UnlistedContractHash)
+                        }
+                    }
+                }
+                KeyTag::Account => runtime::revert(NFTCoreError::InvalidMinter),
+                _ => runtime::revert(NFTCoreError::InvalidKey),
+            }
+        }
```
There is probably a version possible without intermediary booleans, is_whitelisted is_package_whitelisted that would reduce the number of lines but to me the logic becomes more obscure to read for a human, I like to understand just by reading the if statement why a reverts happens without the need to read lines above.

gyroflaw replied:
will get back to you about this at tomorrow

gRoussac replied:
I changed the logic to test the contract package first as maybe this would save some cost for contracts check. Let me know if this looks better to you or let me know otherwise thanks.

gRoussac replied:
I reverted those changes

gRoussac commented on contract/src/main.rs:
```diff
@@ -731,7 +775,10 @@ pub extern "C" fn burn() {
     let token_identifier = utils::get_token_identifier_from_runtime_args(&identifier_mode);
 
-    let caller = utils::get_verified_caller().unwrap_or_revert();
+    let caller: Key = match utils::get_verified_caller().unwrap_or_revert() {
```
get_verified_caller is sending one or two values in an enum so you need to match against them, afaik it would be hard here to make deduced casting, we could put the match into the method but this would complicate the caller handling return.

gRoussac commented on README.md:
```diff
@@ -97,28 +94,13 @@ This modality is an optional installation parameter and will default to the `Mix
 
 1. `Installer`: This mode restricts the ability to mint new NFT tokens only to the installing account of the NFT contract.
 2. `Public`: This mode allows any account to mint NFT tokens.
+3. `Acl`: This mode allows whitelisted account/contract/contract from a package to mint NFT tokens.
```
Right now text is 

`This mode allows whitelisted accounts, contracts or contracts from a package to mint NFT tokens.` as a package hash does not mint directly.

gRoussac committed:
Check pakage first on acl contract

gRoussac committed:
Revert packages changes

gRoussac committed:
Rewrite if statement

gRoussac committed:
true/false is some

gRoussac committed:
Update Casper crates

# Conflicts:
#	client/is_approved_for_all_session/Cargo.toml
#	contract/Cargo.toml
#	tests/src/installer.rs
#	tests/src/mint.rs

gRoussac committed:
Linting

gRoussac committed:
serde-json-wasm version

gRoussac committed:
Merge branch 'feature-1.4' of github.com:casper-ecosystem/cep-78-enhanced-nft into acl

# Conflicts:
#	README.md
#	client/balance_of_session/Cargo.toml
#	client/get_approved_session/Cargo.toml
#	client/is_approved_for_all_session/Cargo.toml
#	client/mint_session/Cargo.toml
#	client/owner_of_session/Cargo.toml
#	client/transfer_session/Cargo.toml
#	client/updated_receipts/Cargo.toml
#	contract/Cargo.toml
#	contract/src/main.rs
#	tests/Cargo.toml

gRoussac committed:
Update test
